test: Delete SOURCE_DATE_EPOCH envvar if already present for robustness

ziyao233 opened the pull request:
Since commit e1f9f3aaa5fc ("Support SOURCE_DATE_EPOCH for reproducible builds (#15)"), calver may pretend current date is value of environment variable "SOURCE_DATE_EPOCH". But when running testsuite in packaging scenarios, the packaging tool, for example, makepkg, may set up its own SOURCE_DATE_EPOCH and break our expected code path to train.

Let's delete the variable from os.environ if it's already present when starting the testsuite, improving its robustness.

ziyao233 committed:
test: Split fixtures into conftest.py

Split test fixtures into conftest.py to clean up and prepare for usage
of pytest_configure() hook. No functional changes involved.

Signed-off-by: Yao Zi <ziyao@disroot.org>

di commented:
I'm confused, the tests that care about `SOURCE_DATE_EPOCH` should use [this fixture](https://github.com/di/calver/blob/e9c8017caf3e3aff214fbb581ee7f366de55f629/tests/test_integration.py#L34-L39) which overwrites `SOURCE_DATE_EPOCH` if it's set in the environment.

Can you share the test failure you're trying to fix with this?

ziyao233 commented:
> I'm confused, the tests that care about `SOURCE_DATE_EPOCH` should use [this fixture](https://github.com/di/calver/blob/e9c8017caf3e3aff214fbb581ee7f366de55f629/tests/test_integration.py#L34-L39) which overwrites `SOURCE_DATE_EPOCH` if it's set in the environment.
> 
> Can you share the test failure you're trying to fix with this?

test_reproducible_build doesn't fail, instead test_version_str and test_version_true do.

```
$ SOURCE_DATE_EPOCH=1 PYTHONPATH=src/ pytest -v -k test_version_str
tests/test_integration.py::test_version_str FAILED                      [100%]

================================== FAILURES ===================================
______________________________ test_version_str _______________________________

dist = <stub(metadata=<stub(version='Thu Jan  1 00:00:01 1970')>)>
keyword = <stub()>, ignore_pkginfo = None

    def test_version_str(dist, keyword, ignore_pkginfo):
        value = "%c"

        calver.integration.version(dist, keyword, value)

>       assert dist.metadata.version == datetime.datetime.utcnow().strftime(value)
E       AssertionError: assert 'Thu Jan  1 00:00:01 1970' == 'Tue Apr 15 15:04:10 2025'
E
E         - Tue Apr 15 15:04:10 2025
E         + Thu Jan  1 00:00:01 1970

tests/test_integration.py:64: AssertionError
```

not only test_reproducible_build() cares about the epoch, but also every test that doesn't rely on PKG-INFO and should succeed. `SOURCE_DATE_EPOCH` has a priority higher than current date[1], so if the variable is set, the output won't match our expectation, which is always constructed from current date.

[1]: https://github.com/di/calver/blob/e9c8017caf3e3aff214fbb581ee7f366de55f629/src/calver/integration.py#L35

di commented:
Got it. I think we should probably move this and other fixtures into [`conftest.py`](https://docs.pytest.org/en/stable/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files) and it should probably use the [`pytest_configure`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.hookspec.pytest_configure) hook instead, something like:

```python
def pytest_configure(config):
    if "SOURCE_DATE_EPOCH" in os.environ:
        del os.environ["SOURCE_DATE_EPOCH"]
```

ziyao233 committed:
test: Delete SOURCE_DATE_EPOCH envvar if already present for robustness

Since commit e1f9f3aaa5fc ("Support SOURCE_DATE_EPOCH for reproducible
builds (#15)"), calver may pretend current date is value of environment
variable "SOURCE_DATE_EPOCH". But when running testsuite in packaging
scenarios, the packaging tool, for example, makepkg, may set up its own
SOURCE_DATE_EPOCH and break our expected code path to train.

Let's delete the variable from os.environ if it's already present when
starting the testsuite, improving its robustness.

Signed-off-by: Yao Zi <ziyao@disroot.org>

ziyao233 commented:
> Got it. I think we should probably move this and other fixtures into [`conftest.py`](https://docs.pytest.org/en/stable/reference/fixtures.html#conftest-py-sharing-fixtures-across-multiple-files) and it should probably use the [`pytest_configure`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.hookspec.pytest_configure) hook instead, something like:
> 
> ```python
> def pytest_configure(config):
>     if "SOURCE_DATE_EPOCH" in os.environ:
>         del os.environ["SOURCE_DATE_EPOCH"]
> ```

Okay, done in the new HEAD. Also added a comment above the hook for explanation.

di committed:
Add full test directory to MANIFEST.in

di committed:
Linting

ziyao233 commented:
Is there anything more I need to do to get this merged? Thanks.
